File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>

long long user_input(void);
int card_length(long long credit_card);
void is_card_valid(long long credit_card);


int main(void)
{
    long long card_number = user_input();
    // int length = card_length(card_number);
    // printf("Card number: %lld\nCard number length: %d\n", card_number, length);
    is_card_valid(card_number);
    return 0;
}


long long user_input(void)
{
    long long card_number;
    /*
    do
    {
        printf("Number: ");
        card_number = get_long_long();
        printf("%d\n", card_length(card_number));
    }
    while(card_length(card_number) < 13 && card_length(card_number) > 16 );
    */
    printf("Number: ");
    card_number = get_long_long();
    return card_number;
}


int card_length(long long credit_card)
{
    int card_length = 0;
    
    for(long long i = credit_card; i > 0; i /= 10)
    {
        card_length++;
    }
    
    return card_length;
}

void is_card_valid(long long credit_card)
{
    int length = card_length(credit_card),
    last_index = length-1,
    sum_non_multiplied = 0,
    sum_multiplied = 0,
    total_sum = 0;
    
    long long card_number_array[length]; 
    
    for (int i = last_index; i >= 0; i--)    // get card number into an array
    {
        card_number_array[i] = credit_card % 10;
        credit_card /= 10;
    }
    
    for (int i = last_index-1; i >= 0; i -= 2)   // sum digits multiplied by 2
    {
        if (((int)card_number_array[i] * 2) >= 10)  // break up two-digit numbers into two digits
        {

            sum_multiplied += ((int)card_number_array[i] * 2) % 10; // ones
            sum_multiplied += ((int)card_number_array[i] * 2) / 10; // tenths
            // printf("number index : %d, number = %d\n", i+1, (int)card_number_array[i]);
            // printf("sum multiplied = %d + %d\n", sum_multiplied, (int)card_number_array[i] * 2);
        }
        else
        {
            sum_multiplied += (int)card_number_array[i] * 2;    // if single digit - sum 
            // printf("number index : %d, number = %d\n", i+1, (int)card_number_array[i]);
            // printf("sum multiplied = %d + %d\n", sum_multiplied, (int)card_number_array[i] * 2);
        }
    }
    
    if (length % 2 == 1)    // if uneven - start at second number
    {
        for (int i = 0; i <= last_index; i += 2) // sum non-multiplied digits
        {
            sum_non_multiplied += (int)card_number_array[i];
            // printf("uneven number index : %d, number = %d\n", i+1, (int)card_number_array[i]);
            // printf("sum nonmultiplied = %d + %d\n", sum_non_multiplied, (int)card_number_array[i] * 2);
        }
    }
    else
    {
        for (int i = 1; i <= last_index; i += 2) // sum non-multiplied digits
        {
            sum_non_multiplied += (int)card_number_array[i];
            // printf("uneven number index : %d, number = %d\n", i+1, (int)card_number_array[i]);
            // printf("sum nonmultiplied = %d + %d\n", sum_non_multiplied, (int)card_number_array[i] * 2);
        }
    }
    
    total_sum = sum_multiplied + sum_non_multiplied; // total sum of multiplied and unmultiplied
    // printf("sum nonmultiplied is %d\n", sum_non_multiplied);
    // printf("sum multiplied is %d\n", sum_multiplied);
    // printf("Total sum: %d\n", total_sum);
    
    if (total_sum % 10 == 0)
    {
        switch (card_number_array[0])   // check what kind of card and if it's valid
        {
            case 3:
            if (card_number_array[1] == 4 || card_number_array[1] == 7)
            {
                printf("AMEX\n");
                break;
            }
            case 4:
            printf("VISA\n");
            break;
            
            case 5:
            if (card_number_array[1] == 1 || card_number_array[1] == 2 
                    || card_number_array[1] == 3 || card_number_array[1] == 4
                        || card_number_array[1] == 5)
                {
                    printf("MASTERCARD\n");
                    break;
                }
                
            default:
            printf("INVALID\n");
            break;
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

File: pset6/credit.py
'''
    Author: Jakub Rupinski
    Excersize from pset6, for CS50 MOOC

    Program takes credit card number from user and says if it's valid or not
'''


def main():

    # test cases
    # VISA
    isValid("4012888888881881")
    # MasterCard
    isValid("5555555555554444")
    # AMEX
    isValid("378282246310005")
    # INVALID
    isValid("6176292929")
    # USER INPUT
    creditCard = getCardNumber()
    isValid(creditCard)


def getCardNumber():

    while True:
        cardNumber = input("Input your credit card number: ")
        # if number length is not valid for VISA, AMEX or MasterCard  - retry
        if (cardNumber.isdigit()) and (len(cardNumber) in [13, 15, 16]):
            break
        print("Not a valid card number, retry...")

    return cardNumber


def isValid(card):

    # keep sums of both multiplied and un-multiplied digits
    sumOfMultiplied = 0
    sumOfNotMultiplied = 0
    sumTotal = 0

    # calculate end point for the for loop
    if len(card) % 2 == 0:
        end = -1
    else:
        end = 0

    # multiply every other digit by 2, starting from second-to-last
    for i in range(len(card) - 2, end, -2):
        digitMultiplied = int(card[i]) * 2
        # add product's digits (not digits themselves) if number has >1 digit
        if digitMultiplied >= 10:
            # add ones
            sumOfMultiplied += int(digitMultiplied % 10)
            # add tens
            sumOfMultiplied += int(digitMultiplied / 10)
        else:
            sumOfMultiplied += digitMultiplied

    # sum non-multiplied digits
    for i in range(abs(end), len(card), 2):
        sumOfNotMultiplied += int(card[i])

    # add both sums and check if card is valid (last digit of total sum == 0)
    sumTotal = sumOfMultiplied + sumOfNotMultiplied
    if sumTotal % 10 == 0:
        getCardBrand(card)
    else:
        print("CHECKSUM INVALID")


def getCardBrand(card):

    firstTwoDigits = int(card[:2])
    # MasterCard numbers all start with 51, 52, 53, 54, or 55
    if firstTwoDigits in [51, 52, 53, 54, 55]:
        print("MASTERCARD")
    # American Express numbers all start with 34 or 37
    elif firstTwoDigits in [34, 37]:
        print("AMEX")
    # Visa numbers all start with 4
    elif int(card[0]) == 4:
        print("VISA")
    else:
        print("INVALID")


if __name__ == "__main__":
    main()
